### PR TITLE
feat: in-house queue system

### PR DESCRIPTION
**Pull request by D4isDAVID**
## Description

Adds an in-house queue system.

The system includes:
- An adaptive card showing where the player is in the queue from start to finish.
- Configurable sub-queues/priority queues.
- Player timeouts:
  - When a player disconnects while in queue, wait a configurable amount of seconds for the player to connect back before removing from the queue.
  - When a player disconnects after being let into the server but before firing the `playerJoining` event (meaning while the player was downloading the server), wait a configurable amount of seconds for the player to connect back before removing.

Some parts of the code may look weird/unnecessary as I've attempted to make it as performant as I can.
I've tested this as much as I can but this still requires further testing for race conditions, edge cases, and other unexpected things that I'm not able to test by myself.

Adaptive card preview:
image

## Checklist

- [x] I have personally loaded this code into an updated Qbox project and checked all of its functionality.
- [x] My code fits the style guidelines.
- [x] My PR fits the contribution guidelines.

### CODE REVIEW

**Review by Manason — CHANGES_REQUESTED** [review]

File: server/queue.lua [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+---@param resource string
+AddEventHandler('onResourceStarting', function(resource)
+    if resource == 'hardcap' then
+        lib.print.info('Preventing hardcap from starting...')
+        CancelEvent()
+    end
+end)
+
+if GetResourceState('hardcap'):find('start') then
+    lib.print.info('Stopping hardcap...')
```
[the remark this answers is not in the document]
Comment (D4isDAVID) [reply server/queue.lua:19]: I don't think so, the average user may be confused on why hardcap was stopped.

File: config/server.lua [hunk 2]
```diff
@@ -136,5 +136,164 @@ return {
...
+            '🕕',
+            '🕘',
+        },
+
+        ---Queue types from most prioritized to least prioritized.
+        ---The first queue without a predicate function will be used as the default.
+        ---If a player doesn't pass any predicates and a queue without a predicate does not exist they will not be let into the server unless player slots are available.
+        ---@type QueueType[]
+        queueTypes = {
+            { name = 'Priority Queue', color = 'good', predicate = function(source) return source % 2 == 0 --[[HasPermission(source, 'admin')]] end },
```
[the remark this answers is not in the document]
Comment (D4isDAVID) [reply config/server.lua:160]: Sorry, this was for testing purposes. Seems like I forgot to remove it.

File: config/server.lua [hunk 3]
```diff
@@ -136,5 +136,164 @@ return {
     
     removeSocietyMoney = function(accountName, payment)
         return exports.qbx_management:RemoveMoney(accountName, payment)
-    end
+    end,
+
+    queue = {
```
[the remark this answers is not in the document]
Comment (D4isDAVID) [reply config/server.lua:141]: I'll move it there since because it's a lot. I wasn't sure at first because all previous configs were fit in the already existing files.

File: server/queue.lua [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    self.positions[license] = self.size
+end
+
+---@param self SubQueue
+---@param license string
+function prototype.dequeue(self, license)
+    local pos = self.positions[license]
+    self.positions[license] = nil
+    self.size -= 1
+    totalSize -= 1
```
[the remark this answers is not in the document]
Comment (D4isDAVID) [reply server/queue.lua:60]: `self.size` is the size of the sub-queue of the player and `totalSize` is the total size of all queues combined. I created `totalSize` instead of a calculate function to not have to calculate the value all the time for all players every second.

File: server/queue.lua [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    self.positions = {}
+    return self
+end
+
+---@param self SubQueue
+---@param license string
+function prototype.enqueue(self, license)
+    self.size += 1
+    totalSize += 1
+    self.positions[license] = self.size
```
[the remark this answers is not in the document]
Comment (D4isDAVID) [reply server/queue.lua:51]: The tradeoff of having license to position is that a less performant loop has to be done when a player is removed from the queue, but the tradeoff of having only position to license is that each time we want to know the position of a player we need to loop. Meaning either we make a less performant loop (since it's done through lua) every time a player disconnects or a more performant loop (I've imagined that `table.remove` would be used since iirc it uses a C loop which theoretically is faster) every second for every player.

I've mapped it license to position because the way I've imagined it is that each player knows their position in the queue and nothing more. I think the only good reason to store position to license/source would be if we need to reveal to a player or an external API what the player in a certain position is, but at the moment the queue does neither, so no reason to store that information currently.

Since the player is updated on their status every second, the position of the player in the queue has to be known every second, but currently it is never asked in what position in line another player is.

**Review by Manason — APPROVED** [review]
